vnext/Microsoft.ReactNative/Views/TouchEventHandler.cpp

```diff
@@ -449,49 +469,76 @@ bool TouchEventHandler::TagFromOriginalSource(
+  // hit box excluding the last character on a line.
+  const auto nextPointer = textPointer.GetPositionAtOffset(1, winrt::LogicalDirection::Forward);
+  auto right = textPointer.VisualParent().Width();
+  if (nextPointer != nullptr) {
+    const auto nextRect = nextPointer.GetCharacterRect(winrt::LogicalDirection::Forward);
+    if (rect.Y == nextRect.Y) {
+      right = nextRect.X;
+    }
+  }
 
-      if (resTag)
-        return resTag;
+  // The character is before the point if the Y-coordinate of the point is
+  // below (greater than) the bottom of the character rect, or if the
+  // Y-coordinate is below (greater than) the top of the character rect and
+  // the X-coordinate is greater than the right side of the character rect:
+  // ┌───────────┐
+  // │     ┌─────┘
+  // └─────┘ (x,y)
+  return point.Y > bottom || (point.Y > rect.Y && point.X > right);
+}
 
-      if (isHit) {
-        tag = el.GetValue(xaml::FrameworkElement::TagProperty()).try_as<winrt::IPropertyValue>();
-        if (tag) {
-          return tag;
-        }
-      }
-    } else if (const auto run = el.try_as<xaml::Documents::Run>()) {
-      const auto start = el.ContentStart();
-      const auto end = el.ContentEnd();
-
-      auto startRect = start.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
-      auto endRect = end.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
-
-      // Swap rectangles in RTL scenarios.
-      if (startRect.X > endRect.X) {
-        const auto tempRect = startRect;
-        startRect = endRect;
-        endRect = tempRect;
-      }
+bool IsCharacterAfter(winrt::TextPointer textPointer, winrt::Point point) {
+  const auto rect = textPointer.GetCharacterRect(winrt::LogicalDirection::Forward);
+  const auto bottom = rect.Y + rect.Height;
+
+  // The character is after the point if the Y-coordinate of the point is above
+  // (less than) the top of the character rect, or if the Y-coordinate is above
+  // (less than) the bottom of the character rect and the X-coordinate is less
+  // than the left side of the character rect:
+  //
+  // (x,y) ┌─────┐
+  // ┌─────┘     │
+  // └───────────┘
+  return point.Y < rect.Y || (point.Y < bottom && point.X < rect.X);
+}
 
-      // Approximate the bounding rect (for now, don't account for text wrapping).
-      if ((startRect.X <= pointerPos.X) && (endRect.X + endRect.Width >= pointerPos.X) &&
-          (startRect.Y <= pointerPos.Y) && (endRect.Y + endRect.Height >= pointerPos.Y)) {
-        isHit = true;
-        return nullptr;
-      }
+winrt::TextPointer TouchEventHandler::GetPositionFromPoint(winrt::TextBlock textBlock, winrt::Point point) {
+  // Since characters in a TextBlock are sorted from top-left to bottom-right,
+  // we can use binary search to find the character whose bounds contains the
+  // pointer point.
+  //
+  // This algorithm currently makes the following assumptions:
+  // 1. Characters on the same line have the same Rect::X value
```

Comment:
should this read same Rect::Y ?
I wonder how this will fare with characters that have large ascenders/descenders, may be something to check

Reply:
Yup.

Reply:
@asklar - Based on the limited amount of testing I did with a few different fonts, it looks like there is never a case where the descenders / ascenders overlap with the next or previous line, so this would not be a problem.

If you think of any fonts where this might not be the case, I'll have a look and see what the impact might be on the algorithm.

I suspect there will not be such a case, because even smaller font sizes on the same line as larger ones have the same character rect Y dimension, so it seems the character rect Y is always the "max"

Reply:
Thanks to the examples provided by @asklar:

ళ ్ ల ీ = ళ్లీ

I was able to confirm that even if the ascender reached beyond the top of the line of text, the Y dimension of all characters in the line are the same (even the special characters with large ascenders).

If you click on the part of the ascender that rises above the top of the character rect, the hit test will *not* pass. I believe this case is somewhat ambiguous as to what the correct behavior is (especially if the ascender overlaps with something else that responds to pointer events), and we should not be concerned about this.

Reply:
thanks for checking @rozele !